crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1220 @@
+            CacheResult::NegativeHit => Ok(None),
+            CacheResult::Miss => self.store.get_marker_value(object_id, version, epoch_id),
+        }
+    }
+
+    fn get_latest_marker(
+        &self,
+        object_id: &ObjectID,
+        epoch_id: EpochId,
+    ) -> SuiResult<Option<(SequenceNumber, MarkerValue)>> {
+        match self.get_latest_marker_value_cache_only(object_id, epoch_id) {
+            CacheResult::Hit((v, marker)) => Ok(Some((v, marker))),
+            CacheResult::NegativeHit => Ok(None),
+            CacheResult::Miss => self.store.get_latest_marker(object_id, epoch_id),
+        }
+    }
+
+    fn get_lock(&self, _obj_ref: ObjectRef, _epoch_id: EpochId) -> SuiLockResult {
+        todo!()
+    }
+
+    fn _get_latest_lock_for_object_id(&self, _object_id: ObjectID) -> SuiResult<ObjectRef> {
+        todo!()
+    }
+
+    fn check_owned_object_locks_exist(&self, _owned_object_refs: &[ObjectRef]) -> SuiResult {
+        todo!()
+    }
+}
+
+impl ExecutionCacheWrite for WritebackCache {
+    #[instrument(level = "trace", skip_all)]
+    fn acquire_transaction_locks<'a>(
+        &'a self,
+        _epoch_id: EpochId,
+        _owned_input_objects: &'a [ObjectRef],
+        _tx_digest: TransactionDigest,
+    ) -> BoxFuture<'a, SuiResult> {
+        todo!()
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    fn write_transaction_outputs(
+        &self,
+        epoch_id: EpochId,
+        tx_outputs: Arc<TransactionOutputs>,
+    ) -> BoxFuture<'_, SuiResult> {
+        let TransactionOutputs {
+            transaction,
+            effects,
+            markers,
+            written,
+            deleted,
+            wrapped,
+            events,
+            ..
+        } = &*tx_outputs;
+
+        // Update all markers
+        for (object_key, marker_value) in markers.iter() {
+            self.dirty
+                .markers
+                .entry((epoch_id, object_key.0))
+                .or_default()
+                .value_mut()
+                .insert(object_key.1, *marker_value);
+        }
+
+        // Write children before parents to ensure that readers do not observe a parent object
+        // before its most recent children are visible.
+        for (object_id, object) in written.iter() {
+            if object.is_child_object() {
+                self.write_object_entry(object_id, object.version(), object.clone().into());
+            }
+        }
+        for (object_id, object) in written.iter() {
+            if !object.is_child_object() {
+                self.write_object_entry(object_id, object.version(), object.clone().into());
+                if object.is_package() {
+                    self.packages
+                        .insert(*object_id, PackageObject::new(object.clone()));
+                }
+            }
+        }
+
+        for ObjectKey(id, version) in deleted.iter() {
+            self.write_object_entry(id, *version, ObjectEntry::Deleted);
+        }
```

Comment:
A newbie question: can the deleted object be a child object? If so, should this also be put before the parent object is updated?

Reply:
I started writing a long reply explaining why this okay, and realized that it is not:

If we delete a child, we must always write a new version of the parent. If we write the new version of the parent first, then a tx can execute, read the child, and not observe the child's tombstone, leading to an incorrect execution.

The fix is simple: We will write all deletes first.

We don't actually have to worry about ordering with the deletes though, because:
- if the parent is deleted no subsequent tx can read either the parent or the child
- if both parent and child are deleted, and we write the child tombstone first, it will not be observed by a reader reading the prior version of the parent due to MVCC

Reply:
fixed and added test that finds the bug